### PR TITLE
Support for all three modes (privacy, integrity, authentication)

### PR DESCRIPTION
**Pull request by bdrosen96**

### CODE REVIEW

File: src/rpc/RpcChannel.cpp [hunk 1]
```diff
@@ -167,19 +234,33 @@ std::string RpcChannelImpl::saslEvaluateToken(RpcSaslProto & response, bool serv
...
-            THROW(AccessControlException, "Client generated spurious response");
+            if (!saslClient || !saslClient->isPrivate())
+                THROW(AccessControlException, "Client generated spurious response");
         }
     }
 
     return token;
 }
 
+void spin(const char* strname) {
```
Comment (jettify) [inline src/rpc/RpcChannel.cpp:245]: is this debugging code?

Comment (bdrosen96) [reply src/rpc/RpcChannel.cpp:245]: yes, I will remove it once tokens are working

File: src/rpc/RpcChannel.cpp [hunk 2]
```diff
@@ -51,6 +52,70 @@ using namespace google::protobuf::io;
...
+        rpcHeader.set_rpcop(RpcRequestHeaderProto_OperationProto_RPC_FINAL_PACKET);
+        RpcContentWrapper wrapper(&rpcHeader, &msg);
+        totalLen = wrapper.getLength();
+        buffer.writeBigEndian(totalLen);
+        wrapper.writeTo(buffer);
+
+        int length = buffer.getDataSize(0);
+        std::string outdata;
+        outdata.resize(length);
+        memcpy(&outdata[0], buffer.getBuffer(0), length);
```
Comment (izeigerman) [inline src/rpc/RpcChannel.cpp:86]: bad idea to use low-level functions for high-level C++ objects.:
If `buffer.getBuffer(0)` returns `char *`:

```
const char * buf = buffer.getBuffer(0)
std::string outdata(buf, buf + length)
```

or if not:

```
const char * buf = (const char *)buffer.getBuffer(0) // better use reinterpret_cast or static_cast.
std::string outdata(buf, buf + length)
```

File: src/rpc/RpcChannel.cpp [hunk 3]
```diff
@@ -526,7 +610,16 @@ void RpcChannelImpl::sendRequest(RpcRemoteCallPtr remote) {
     WriteBuffer buffer;
     assert(true == available);
     remote->serialize(key.getProtocol(), buffer);
-    sock->writeFully(buffer.getBuffer(0), buffer.getDataSize(0),
+    std::string data;
+    int length = buffer.getDataSize(0);
+    data.resize(length);
+    memcpy(&data[0], buffer.getBuffer(0), length);
```
Comment (izeigerman) [inline src/rpc/RpcChannel.cpp]: same as above

File: src/rpc/RpcChannel.cpp [hunk 4]
```diff
@@ -697,7 +796,16 @@ void RpcChannelImpl::sendConnectionContent(const RpcAuth & auth) {
     int size = wrapper.getLength();
     buffer.writeBigEndian(size);
     wrapper.writeTo(buffer);
-    sock->writeFully(buffer.getBuffer(0), buffer.getDataSize(0),
+    int length = buffer.getDataSize(0);
+    data.resize(length);
+    memcpy(&data[0], buffer.getBuffer(0), length);
```
Comment (izeigerman) [inline src/rpc/RpcChannel.cpp]: same

File: src/rpc/SaslClient.cpp [hunk 5]
```diff
@@ -152,15 +178,85 @@ std::string SaslClient::evaluateChallenge(const std::string & challenge) {
...
     }
 
     return retval;
 }
 
+std::string SaslClient::encode(const char *input, size_t input_len) {
+    std::string result;
+    if ((!privacy && !integrity) || (!complete)) {
+        result.resize(input_len);
+        memcpy(&result[0], input, input_len);
```
Comment (izeigerman) [inline src/rpc/SaslClient.cpp]: same

File: src/rpc/SaslClient.cpp [hunk 6]
```diff
@@ -152,15 +178,85 @@ std::string SaslClient::evaluateChallenge(const std::string & challenge) {
...
+        return result;
+    }
+    spin("/tmp/brett");
+
+    char *output=NULL;
+    size_t output_len;
```
Comment (izeigerman) [inline src/rpc/SaslClient.cpp:214]: uninitialized integer

File: src/rpc/SaslClient.cpp [hunk 7]
```diff
@@ -152,15 +178,85 @@ std::string SaslClient::evaluateChallenge(const std::string & challenge) {
...
+    }
+    printf("before %d %d\n", input_len, result.length());
+    return result;
+}
+
+std::string  SaslClient::decode(const char *input, size_t input_len) {
+    std::string result;
+    if ((!privacy && !integrity) || (!complete)) {
+        result.resize(input_len);
+        memcpy(&result[0], input, input_len);
```
Comment (izeigerman) [inline src/rpc/SaslClient.cpp]: same as above

File: src/rpc/SaslClient.cpp [hunk 8]
```diff
@@ -152,15 +178,85 @@ std::string SaslClient::evaluateChallenge(const std::string & challenge) {
...
+    }
+    char *output=NULL;
+    size_t output_len;
+    int rc = gsasl_decode(session, input, input_len, &output, &output_len);
+    if (rc != GSASL_OK) {
+        THROW(AccessControlException, "Failed to decode wrapped data: %s", gsasl_strerror(rc));
+    }
+    if (output_len) {
+        result.resize(output_len);
+        memcpy(&result[0], output, output_len);
```
Comment (izeigerman) [inline src/rpc/SaslClient.cpp]: same

File: src/rpc/SaslClient.cpp [hunk 9]
```diff
@@ -39,9 +39,11 @@ namespace Internal {
 
 SaslClient::SaslClient(const RpcSaslProto_SaslAuth & auth, const Token & token,
                        const std::string & principal) :
-    complete(false) {
+    complete(false), theAuth(auth), theToken(token), thePrincipal(principal) {
     int rc;
     ctx = NULL;
+    privacy = false;
+    integrity = false;
```
Comment (izeigerman) [inline src/rpc/SaslClient.cpp]: why don't you initialize this in initialization list instead?

File: src/rpc/RpcChannel.cpp [hunk 10]
```diff
@@ -51,6 +52,70 @@ using namespace google::protobuf::io;
...
+    }
+
+    std::string wrap(std::string data) {
+        if (!saslClient->isPrivate() && !saslClient->isIntegrity())
+            return data;
+
+        std::string rawdata = saslClient->encode(data.c_str(), data.length());
+        msg.set_state(RpcSaslProto_SaslState_WRAP);
+        msg.set_token(rawdata);
+        int totalLen;
```
Comment (izeigerman) [inline src/rpc/RpcChannel.cpp:70]: uninitialized integer.

File: src/rpc/RpcChannel.cpp [hunk 11]
```diff
@@ -657,7 +750,12 @@ void RpcChannelImpl::sendConnectionHeader(const RpcAuth &auth) {
     buffer.write(static_cast<char>(RPC_HEADER_VERSION));
     buffer.write(static_cast<char>(0));  //for future feature
     buffer.write(static_cast<char>(auth.getProtocol()));
-    sock->writeFully(buffer.getBuffer(0), buffer.getDataSize(0),
+    std::string data;
+    int length = buffer.getDataSize(0);
+    data.resize(length);
+    memcpy(&data[0], buffer.getBuffer(0), length);
```
Comment (izeigerman) [inline src/rpc/RpcChannel.cpp]: same stuff with string initialization.

**Comment by izeigerman** [comment]
Btw if you're using std::string only as a raw/binary buffer container why don't you use `std::vector<std::unit8_t>` instead? std::string is not very applicable for this kind of operations.

File: src/rpc/RpcChannel.cpp (hunk 2 above)
Comment (izeigerman) [reply src/rpc/RpcChannel.cpp:86]: Also see my comment about std::vector<char> instead of std::string.

Comment (bdrosen96) [reply src/rpc/RpcChannel.cpp:86]: I did it this way as this is consistent with the rest of the code base in this file.

File: src/rpc/SaslClient.cpp (hunk 9 above)
Comment (bdrosen96) [reply src/rpc/SaslClient.cpp]: mostly because I don't like that style of initialization for simple objects.

File: src/rpc/SaslClient.cpp (hunk 6 above)
Comment (bdrosen96) [reply src/rpc/SaslClient.cpp:214]: It is set by gsasl_encode as an output.

File: src/rpc/RpcChannel.cpp (hunk 10 above)
Comment (bdrosen96) [reply src/rpc/RpcChannel.cpp:70]: It is set below after data is read. No point to initialize it before then as it is not used.

File: src/rpc/RpcChannel.cpp [hunk 12]
```diff
@@ -772,10 +880,11 @@ static exception_ptr HandlerRpcResponseException(exception_ptr e) {
 void RpcChannelImpl::readOneResponse(bool writeLock) {
```
Comment (dalazx) [inline src/rpc/RpcChannel.cpp:880]: that's a long function. should we try to extract SASL related logic into a separate method?

**Comment by bdrosen96** [comment]
std string is used in a number of places already so I tried to be consistent with using it the same way. It is also easier to use with the debugger and can be cast to c types properly.

File: src/rpc/RpcChannel.cpp (hunk 12 above)
Comment (bdrosen96) [reply src/rpc/RpcChannel.cpp:880]: No so interested in refactoring code that already exists beyond getting this working.

File: src/rpc/RpcChannel.cpp (hunk 2 above)
Comment (izeigerman) [reply src/rpc/RpcChannel.cpp:86]: hm. got it. removing the rest of comments.

File: src/rpc/RpcChannel.cpp (hunk 10 above)
Comment (izeigerman) [reply src/rpc/RpcChannel.cpp:70]: ah, I see. Maybe it's good idea to declare it in-place.

File: src/rpc/SaslClient.cpp (hunk 9 above)
Comment (izeigerman) [reply src/rpc/SaslClient.cpp]: It leads to double initialization. At first the variable is created and initalized, than in constructor body you assign a new value to it. It's not so bad for boolean values but not very nice for complex objects.

**Comment by jettify** [comment]
lets merge this

**Pull request merged**